datasetinsights/io/downloader/unity_simulation.py

```
@@ -0,0 +1,120 @@
+                unity simulation the dataset should be downloaded from.
+
+                The expected source-uri follows these patterns
+
+                usim://access-token@project-id/run-execution-id
+                or
+                usim://project-id/run-execution-id
+
+
+            output: This is the path to the directory
+                where the download will store the dataset.
+
+            include_binary: Whether to download binary files
+                such as images or LIDAR point
+                clouds. This flag applies to Datasets where metadata
+                (e.g. annotation json, dataset catalog, ...)
+                 can be separated from binary files.
+
+        """
+        self.parse_source_uri(source_uri)
+        manifest_file = os.path.join(output, f"{self.run_execution_id}.csv")
+        if self.access_token:
+            manifest_file = download_manifest(
+                self.run_execution_id,
+                manifest_file,
+                self.access_token,
+                project_id=self.project_id,
+            )
+        else:
+            logger.info(
+                f"No auth token is provided. Assuming you already have "
+                f"a manifest file located in {manifest_file}"
+            )
+
+        dl_worker = Downloader(manifest_file, output)
+        dl_worker.download_references()
+        dl_worker.download_metrics()
+        dl_worker.download_captures()
+        if include_binary:
+            dl_worker.download_binary_files()
+
+    def parse_source_uri(self, source_uri):
+        """
+
+        Args:
+            source_uri: Parses source-uri in the following format
+            usim://access-token@project-id/run-execution-id
+            or
+            usim://project-id/run-execution-id
+
+        """
+        pattern = re.compile(self.SOURCE_URI_PATTERN)
+        if pattern.findall(source_uri):
+            (access_token, project_id, run_execution_id,) = pattern.findall(
```

What if `pattern.findall` found multiple matches? Will it be better if we raise exceptions?

---

I don't understand what you mean. Could you elaborate a bit?

---

If users has input `http://domain.file usim://uuid/run` as a single string, it will actually silently fail to download the second one. This in fact should not be a valid case as our implementation should support only single source uri.

---

Are you thinking of the scenario when a user imports this class and calls parse_source_uri(source_uri) directly? The example you gave above should be stopped by the cli.

---

What is the input being given to?